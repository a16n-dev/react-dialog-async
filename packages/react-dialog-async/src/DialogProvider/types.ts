import type { PropsWithChildren } from 'react';

export interface DialogProviderProps extends PropsWithChildren {
  /**
   * The default delay in milliseconds to wait before unmounting a dialog after it's closed.
   */
  defaultUnmountDelayInMs?: number;
}
