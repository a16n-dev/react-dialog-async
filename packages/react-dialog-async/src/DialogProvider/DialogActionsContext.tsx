import { createContext } from 'react';
import type { AsyncDialogComponent } from '../types.js';

export interface DialogActionsContextValue {
  show: (
    dialogId: string,
    hash: number,
    dialog: AsyncDialogComponent<any, any>,
    data: unknown,
    unmountDelay?: number,
  ) => Promise<any>;
  hide: (dialogId: string, data?: any) => void;
  updateData: (dialogId: string, data: unknown) => void;
}

export const DialogActionsContext =
  createContext<DialogActionsContextValue | null>(null);
