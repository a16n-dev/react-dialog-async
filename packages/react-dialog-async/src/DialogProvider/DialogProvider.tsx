import React, { useContext, useEffect, useMemo, useRef } from 'react';
import { PropsWithChildren, useCallback, useState } from 'react';
import DialogContext, { dialogContextState } from '../DialogContext';
import { DialogComponent } from '../types';
import DialogStateContext, { dialogsStateData } from '../DialogStateContext';
import useRenderDialogs from '../useRenderDialogs';

interface DialogProviderProps extends PropsWithChildren {
  /**
   * The default delay in milliseconds to wait before unmounting a dialog after it's closed.
   */
  defaultUnmountDelayInMs?: number;
  /**
   * If provided, this function will be called when useDialogLazy is mounted.
   * This lets you call the preload function at some point before the lazy component is required
   * Consult the docs for some sensible functions to provide here for different environments
   */
  lazyLoaderFn?: (preload: () => Promise<void>) => Promise<void>;
}

const DialogProvider = ({
  defaultUnmountDelayInMs,
  lazyLoaderFn,
  children,
}: DialogProviderProps) => {
  // This ref tracks timers for unmount dialogs after they're closed
  const unmountDelayTimeoutRefs = useRef<{ [key: string]: any }>({});

  const [dialogState, setDialogState] = useState<dialogsStateData>({});
  const [usingOutlet, setUsingOutlet] = useState(false);

  /**
   * Force closes the dialog with the given id.
   */
  const hide = useCallback((id: string) => {
    setDialogState((state) => {
      if (!state[id]) return state;

      if (!state[id].open) return state; // don't do anything if the dialog is already closed

      state[id].resolve?.(undefined);

      if (state[id].unmountDelay) {
        if (unmountDelayTimeoutRefs.current[id] !== undefined) {
          clearTimeout(unmountDelayTimeoutRefs.current[id]);
        }

        // start the delay
        unmountDelayTimeoutRefs.current[id] = setTimeout(() => {
          setDialogState((state) => removeKey(state, id));
        }, state[id].unmountDelay);

        return {
          ...state,
          [id]: {
            open: false,
            dialog: state[id].dialog,
            data: state[id].data,
          },
        };
      }
      return removeKey(state, id);
    });
  }, []);

  const show = useCallback(
    (
      id: string,
      hash: number,
      dialog: DialogComponent<any, any>,
      data: unknown,
      unmountDelay?: number,
    ): Promise<unknown> => {
      return new Promise((resolve) => {
        if (unmountDelayTimeoutRefs.current[id] !== undefined) {
          clearTimeout(unmountDelayTimeoutRefs.current[id]);
        }

        const resolveFn = (value: any) => {
          resolve?.(value);
          hide(id);
        };

        setDialogState((state) => {
          if (state[id]?.open) {
            resolve(undefined);
            return state;
          }

          return {
            ...state,
            [id]: {
              dialog: dialog,
              open: true,
              hash,
              data,
              resolve: resolveFn,
              unmountDelay: unmountDelay ?? defaultUnmountDelayInMs,
            },
          };
        });
      });
    },
    [hide, defaultUnmountDelayInMs],
  );

  /**
   * Updates the data of the given dialog
   */
  const updateData = useCallback((id: string, data: unknown): void => {
    setDialogState((state) => {
      if (state[id]) {
        return {
          ...state,
          [id]: { ...state[id], data },
        };
      }
      return state;
    });
  }, []);

  useEffect(() => {
    return () => {
      Object.values(unmountDelayTimeoutRefs.current).forEach(clearTimeout);
    };
  }, []);

  /**
   * To prevent unnecessary re-renders, be careful to ensure that this state has
   * no transitive dependencies to `dialogState`
   */
  const ctx: dialogContextState = useMemo(
    () => ({
      show,
      hide,
      updateData,
      lazyLoaderFn,
    }),
    [show, hide, updateData],
  );

  return (
    <DialogStateContext.Provider
      value={{ dialogs: dialogState, setIsUsingOutlet: setUsingOutlet }}
    >
      <DialogContext.Provider value={ctx}>
        {children}
        {!usingOutlet && <InternalDialogOutlet />}
      </DialogContext.Provider>
    </DialogStateContext.Provider>
  );
};

export default DialogProvider;

function removeKey<
  S extends string | number | symbol,
  T extends Record<S, unknown>,
>(data: T, key: S): Omit<T, S> {
  const { [key]: _, ...rest } = data;
  return rest;
}

const InternalDialogOutlet = () => {
  const dialogState = useContext(DialogStateContext);

  if (!dialogState) {
    throw new Error(
      'Dialog context not found. You likely forgot to wrap your app in a <DialogProvider/> (https://react-dialog-async.a16n.dev/installation)',
    );
  }

  const dialogComponents = useRenderDialogs(dialogState.dialogs);

  if (process.env.NODE_ENV !== 'production' && dialogComponents.length > 0) {
    console.warn(
      'Rendering a dialog without a <DialogOutlet/>. Please include a <DialogOutlet/> as a child of <DialogProvider/> to ensure dialogs are rendered within the correct contexts - this will be required in the next major version of react-dialog-async. See https://react-dialog-async.a16n.dev/API/dialog-outlet for more details. This warning is only present in development',
    );
  }

  return <>{dialogComponents}</>;
};
