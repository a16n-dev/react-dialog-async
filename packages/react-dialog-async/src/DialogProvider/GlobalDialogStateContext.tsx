import { createContext } from 'react';
import type { AsyncDialogComponent } from '../types.js';

export type dialogsStateData = Record<
  string,
  {
    dialog: AsyncDialogComponent<any, any>;
    hash: number;
    data: unknown;
    open: boolean;
    resolve?: (value?: unknown) => void;
    unmountDelay?: number;
  }
>;

export type GlobalDialogStateContextValue = {
  setIsUsingOutlet: (value: boolean) => void;
  dialogs: dialogsStateData;
};

export const GlobalDialogStateContext =
  createContext<GlobalDialogStateContextValue | null>(null);
