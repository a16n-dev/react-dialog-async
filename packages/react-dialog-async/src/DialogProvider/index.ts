export { default as DialogProvider } from './DialogProvider';
export { type DialogProviderProps } from './types';
