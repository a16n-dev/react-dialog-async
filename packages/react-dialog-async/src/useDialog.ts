import { useContext, useEffect, useId, useMemo } from 'react';
import DialogContext from './DialogContext';
import { hashComponent } from './utils';
import { DialogComponent, useDialogOptions, useDialogReturn } from './types';

function useDialog<D, R, DE extends D | undefined, M extends boolean = false>(
  component: DialogComponent<D, R>,
  options: useDialogOptions<D, DE, M> = {},
): useDialogReturn<D, R, DE> {
  const id = useId();

  const key = useMemo(() => hashComponent(component), [component]);

  const ctx = useContext(DialogContext);

  useEffect(() => {
    const unregister = ctx.register(id, key, component);
    return () => {
      unregister();
    };
  }, [id]);

  const show = async (data?: D): Promise<R | undefined> => {
    return ctx.show(id, data ?? options.defaultData);
  };

  const hide = () => {
    return ctx.hide(id);
  };

  const updateData = (data: D) => {
    return ctx.updateData(id, data);
  };

  return {
    show,
    hide,
    updateData,
    open: show,
    close: hide,
  };
}

export default useDialog;
