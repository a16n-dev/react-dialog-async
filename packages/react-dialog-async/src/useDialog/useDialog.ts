import {
  useCallback,
  useContext,
  useEffect,
  useId,
  useMemo,
  useRef,
} from 'react';
import type { AsyncDialogComponent } from '../types.js';
import type { useDialogOptions, useDialogReturn } from './types.js';
import { DialogActionsContext } from '../DialogProvider/DialogActionsContext.js';

export function useDialog<D, R, DE extends D | undefined>(
  component: AsyncDialogComponent<D, R>,
  options?: useDialogOptions<D, DE>,
): useDialogReturn<D, R, DE> {
  const idCount = useRef(0);
  const internalId = useId();

  const id = useMemo(() => {
    if (options?.customKey !== undefined) {
      return options.customKey;
    }

    return internalId;
  }, [internalId, component, options?.customKey]);

  const ctx = useContext(DialogActionsContext);

  if (!ctx) {
    throw new Error(
      'Dialog context not found. You likely forgot to wrap your app in a <DialogProvider/> (https://react-dialog-async.a16n.dev/installation)',
    );
  }

  useEffect(() => {
    return () => {
      if (options?.hideOnHookUnmount !== false) {
        ctx.hide(id);
      }
    };
  }, [id, options?.hideOnHookUnmount]);

  const show = useCallback(
    async (data?: D): Promise<R | undefined> => {
      return ctx.show(
        id,
        idCount.current++,
        component,
        data ?? options?.defaultData,
        options?.unmountDelayInMs,
      );
    },
    [id, component, options?.defaultData, options?.unmountDelayInMs],
  );

  const hide = () => {
    return ctx.hide(id);
  };

  const updateData = (data: D) => {
    return ctx.updateData(id, data);
  };

  return {
    show,
    hide,
    updateData,
    open: show,
    close: hide,
  };
}
