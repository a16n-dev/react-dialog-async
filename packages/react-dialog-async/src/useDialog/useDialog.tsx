import { useCallback, useContext, useEffect, useId, useRef } from 'react';
import type { AsyncDialogComponent } from '../types.js';
import type { useDialogOptions, useDialogReturn } from './types.js';
import { DialogActionsContext } from '../context/DialogActionsContext.js';

/**
 * @category Hooks
 */
export function useDialog<D, R, DE extends D | undefined>(
  component: AsyncDialogComponent<D, R>,
  options?: useDialogOptions<D, DE>,
): useDialogReturn<D, R, DE> {
  const idCount = useRef(0);
  const id = useId();

  const ctx = useContext(DialogActionsContext);

  if (!ctx) {
    throw new Error(
      'Dialog context not found. You likely forgot to wrap your app in a <DialogProvider/> (https://react-dialog-async.a16n.dev/installation)',
    );
  }

  useEffect(() => {
    return () => {
      if (options?.persistOnUnmount !== true) {
        ctx.hide(id);
      }
    };
  }, [id, options?.persistOnUnmount]);

  const open = useCallback(
    async (data?: D): Promise<R | undefined> => {
      return ctx.show(
        id,
        idCount.current++,
        component,
        data ?? options?.defaultData,
        options?.unmountDelayInMs,
      );
    },
    [id, component, options?.defaultData, options?.unmountDelayInMs],
  );

  const close = () => {
    return ctx.hide(id);
  };

  const updateData = (data: D) => {
    return ctx.updateData(id, data);
  };

  return {
    updateData,
    open,
    close,
  };
}
