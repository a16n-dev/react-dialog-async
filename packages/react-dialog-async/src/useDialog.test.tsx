import { expect, test } from 'vitest';
import { render } from '@testing-library/react';
import DialogProvider from './DialogProvider/DialogProvider';

import useDialog from './useDialog';

const TestDialog = () => <div>Hello World!</div>;

test('can be called without error', () => {
  const TestComponent = () => {
    useDialog(TestDialog);

    return null;
  };
  const result = render(
    <DialogProvider>
      <TestComponent />
    </DialogProvider>,
  );

  expect(result.asFragment()).toMatchSnapshot();
});
