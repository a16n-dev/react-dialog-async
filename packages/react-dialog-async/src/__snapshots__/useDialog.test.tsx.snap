// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`can be called without error 1`] = `<DocumentFragment />`;
